sources/packages/api/src/route-controllers/education-program-offering/education-program-offering.controller.service.ts

```diff
@@ -108,4 +134,152 @@ export class EducationProgramOfferingControllerService {
       description: getOfferingNameAndPeriod(offering),
     }));
   }
+
+  /**
+   * Creates the offering model to be validated and saved using the DTO received.
+   * @param institutionId institution id.
+   * @param locationId location id.
+   **The authorization to validated if the location belongs to the institution
+   **is not expected to be performed by this method.
+   * @param programId program id.
+   * @param payload information to generate the model to perform the offering
+   * validation and persistence.
+   * @returns offering model to be validated and saved.
+   */
+  async mapOfferingAPIInDTOTOSaveOfferingModel(
+    institutionId: number,
+    locationId: number,
+    programId: number,
+    payload: EducationProgramOfferingAPIInDTO,
+  ): Promise<SaveOfferingModel> {
+    // Program information required to perform the offering validations.
+    const program = await this.programService.getEducationProgramDetails(
+      programId,
+      institutionId,
+    );
+    if (!program) {
+      throw new NotFoundException(
+        "Program to create the offering not found for the institution.",
+      );
+    }
+    return {
+      ...payload,
+      locationId,
+      studyBreaks: payload.breaksAndWeeks?.studyBreaks,
+      programContext: program,
+    };
+  }
+
+  /**
+   * Verify if all CSV validations were performed with success and throw
+   * a BadRequestException in case of some failure.
+   * @param csvValidations validations to be verified.
+   */
+  assertCSVValidationsAreValid(csvValidations: OfferingCSVValidationResult[]) {
+    const csvValidationsErrors = csvValidations.filter(
+      (csvValidation) => csvValidation.errors.length,
+    );
+    if (csvValidationsErrors.length) {
+      // At least one error was detected and the CSV must be fixed.
+      const validationResults: OfferingBulkInsertValidationResultAPIOutDTO[] =
+        csvValidationsErrors.map((validation) => ({
+          recordIndex: validation.index,
```

Comment:
is this index different from `.map((x,index)`

Reply:
For CSV and Offering model validation they will match. This index is also carried to the insert operation where the inserts will happen in parallel and the order will not be the same.
For this specific one does not matter.